### DISCUSSION TITLE
Is my problem eglot or clangd?

### DISCUSSION THREAD

**Discussion by vincentjgoh** [joaotavora/eglot#1010] [Q&A] [answered]
So I'm struggling with huge stalls (30-60s, sometimes more) from using eglot + clangd. I thought I had it figured out, but I guess I don't.

I'm using eglot basically stock out of the box; I haven't seen anything that seems worthwhile configuring. I've turned off hover-info, and I've tried to disable flymake (it spits out mostly incorrect errors, honestly).

My project is huge. It's an Unreal Engine project, and there are a lot of files. The compile_commands.json is 700+MB (most of it is flags, though, and a lot of them are duplicates), but it mostly works. clangd is currently taking up about 6GB of memory (1GB working) on my Windows 10 machine.

But if I open up the events buffer, there's almost always something compiling. File after file after file, even files that haven't changed since I last saw them go by in the events buffer. I added a .clangd to my dev directory (and I believe it's being picked up; clangd uses config files by default, as near as I can tell, and I've used --check=<file> on the command line to make sure there's nothing bad in it).

Because there are so many files compiling, I thought maybe I could get around that by adding
```
CompileFlags:
    Remove: -W*

Diagnostics:
    Suppress: '*'

Index:
    Background: Skip
    External:
	    File: d:\p4v\Dev\clangd.dex
```

to my .clangd file. No real change, still huge hitches all the time. FWIW, the clangd.dex file is about 315MB. I have NO idea if clangd is using it properly (if I'm honest, the clangd documentation is a nightmare; it's so bare-bones and is absolutely not helpful in a case like this).

Edit:
```
[stderr] I[09:28:34.687] --> reply:textDocument/signatureHelp(115) 5260 ms
[server-notification] Wed Aug 17 09:28:46 2022:
(:jsonrpc "2.0" :method "textDocument/publishDiagnostics" :params
		  (:diagnostics
		   []
		   :uri "file:///d:/p4v/Dev/Source/Team/Team.cpp" :version 0))
[stderr] I[09:28:46.179] --> textDocument/publishDiagnostics
```
(Does this show that the .clangd is being ignored? Diagnostics are supposed to be off.)

So I suppose the question is whether this is an issue I'm having with clangd being pathological, or eglot making unnecessary requests. I love the functionality, but having the whole editor stall when I'm halfway through a line, or right after I save or any other number of somewhat random times is completely untenable, and I can't believe that this is the experience everyone else is having. I've had emacs stall while I was typing out a COMMENT in a header file. That can't possibly be normal behaviour.

If anyone has any help, I'd really appreciate it.

**Comment by vincentjgoh** [comment]
Oh, also, I think it might be worth having an option to pause any requests while a compile is happening. I've launched a long build only to see clangd chewing through files in the background, which seems counter-productive.

**Reply by joaotavora** [reply]
Let's take measurements first and take measures after.

**Comment by joaotavora** [comment]
vincentjgoh ***@***.***> writes:

> So I suppose the question is whether this is an issue I'm having with
> clangd being pathological, or eglot making unnecessary requests.

A poor-man's way to tell what's eating up your CPU and your time is to
glance at 'top's output on another terminal window.  If you see the
'emacs' program peaking over `clangd`, then something in Emacs --
possibly but not necessarily -- Eglot, is the culprit.

Yet another way is to use M-x profiler-start, M-x profiler-stop and M-x
profiler-report and post the expanded relevant portion of the results in
this thread.

> I love the functionality, but having the whole editor stall when I'm
> halfway through a line, or right after I save or any other number of
> somewhat random times is completely untenable, and I can't believe
> that this is the experience everyone else is having. I've had emacs
> stall while I was typing out a COMMENT in a header file. That can't
> possibly be normal behaviour.

No, it isn't.  But don't assume comments aren't somehow relevant to the
LSP server, they are.  In particular, clangd can work with the
clang-tidy tool, which is generally aware of rule
activation/deactivation directives that are encoded in comments.

I use Eglot and clangd very successfully, though not with "monstruous"
projects.  However, including header files with lots of templates does
mean that my compilation times are non-trivial.  But my editing
experience isn't extremely degraded.  And I don't turn off any default
feature of Eglot, though I do happen to use a particularly lean and
vanilla configuration.

So your problem could come down to a number of things:

* Eglot making too many requests because you've set some variable like
  eglot-send-changes-idle-time or eldoc-idle-delay too low.

* You're using some other eager thing like company-mode and it also
  makes "idle-time" requests in the (pseudo-)background controled by
  company-idle-delay.

* The printing of log information to *EGLOT (project/c++-mode) events*
  is costing a lot of time and memory.  In this case you might have luck
  setting eglot-events-buffer-size to 0.

* You're using some other greedy extension in Emacs, be sure to conduct
  any tests using a controlled `emacs -Q` run.

As usual, the only way to tell is to conduct measurements, so be sure to
send some `M-x profiler-report` output this way.

FWIW's here's a typical profile from 2-3 minutes editing in a C++
project.  As you can see Eglot's code isn't even on the map.  But
c++-mode's fontification code is, though not in a show-stopping way.
c++-mode is notoriously complex and slow (and buggy), maybe someday
we'll get to replace it with something better.

João

**Reply by joaotavora** [reply]
Forgot to post the profile, here is some of it:
```
        1261  45% - redisplay_internal (C function)
        1136  40%  - jit-lock-function
        1133  40%   - jit-lock-fontify-now
        1133  40%    - jit-lock--run-functions
        1133  40%     - run-hook-wrapped
        1133  40%      - #<compiled -0x156e66ed23bd7283>
        1133  40%       - font-lock-fontify-region
        1133  40%        - c-font-lock-fontify-region
        1125  40%         - font-lock-default-fontify-region
        1122  40%          - font-lock-fontify-keywords-region
         863  31%           - c-font-lock-declarations
         852  30%            - c-find-decl-spots
         421  15%             - c-bs-at-toplevel-p
         417  15%              - c-brace-stack-at
         402  14%               - c-update-brace-stack
                ...
          82   2%  + eval
          13   0%    file-remote-p
          12   0%  + #<compiled -0x84ff12d031f8270>
           3   0%  + redisplay--pre-redisplay-functions
         871  31% - command-execute
         871  31%  - call-interactively
         727  26%   - byte-code
         727  26%    - read-extended-command
         727  26%     - read-extended-command-1
         727  26%      - completing-read
         727  26%       - completing-read-default
         726  26%        - read-from-minibuffer
         538  19%         - icomplete-post-command-hook
         538  19%          - icomplete-exhibit
         530  19%           - icomplete-completions
         530  19%            - icomplete--sorted-completions
```

By the way, the `command-execute` part isn't very relevant: the profiler was only measuring the effort it took my completion system (`icomplete`) to display the options from which I chose `M-x profiler-stop`.  That this shows up in the profile at all is a relatively good sign, it means my Emacs was mostly idle and waiting for user input for the 2-3 minutes that this profiling run took.

**Reply by vincentjgoh** [reply]
> vincentjgoh ***@***.***> writes:
>> So I suppose the question is whether this is an issue I'm having with clangd being pathological, or eglot making unnecessary requests.

> A poor-man's way to tell what's eating up your CPU and your time is to glance at 'top's output on another terminal window. If you see the 'emacs' program peaking over `clangd`, then something in Emacs -- possibly but not necessarily -- Eglot, is the culprit. Yet another way is to use M-x profiler-start, M-x profiler-stop and M-x profiler-report and post the expanded relevant portion of the results in this thread.

I'll post the results in a different post so as not to clog this one up.

>> I love the functionality, but having the whole editor stall when I'm halfway through a line, or right after I save or any other number of somewhat random times is completely untenable, and I can't believe that this is the experience everyone else is having. I've had emacs stall while I was typing out a COMMENT in a header file. That can't possibly be normal behaviour.

> No, it isn't. But don't assume comments aren't somehow relevant to the LSP server, they are. In particular, clangd can work with the clang-tidy tool, which is generally aware of rule activation/deactivation directives that are encoded in comments. I use Eglot and clangd

Is that on by default? If it's not, I'm not using it. (Once again, this is an obnoxious problem with the whole clangd toolchain. I refer back to the documentation several times a day at this point, and trying to figure out what's enabled and what's not is a huge problem. The examples are meagre, not necessarily platform agnostic, and clangd itself doesn't give much insight into what it's doing, even when you turn on really verbose diagnostics)

>very successfully, though not with "monstruous" projects. However, including header files with lots of templates does mean that my compilation times are non-trivial. But my editing experience isn't extremely degraded. And I don't turn off any default feature of Eglot, though I do happen to use a particularly lean and vanilla configuration. So your problem could come down to a number of things: * Eglot making too many requests because you've set some variable like eglot-send-changes-idle-time or eldoc-idle-delay too low. * You're using some other eager thing like company-mode and it also makes "idle-time" requests in the (pseudo-)background controled by company-idle-delay. * The printing of log information to *EGLOT (project/c++-mode) events* is costing a lot of time and memory. In this case you might have luck setting eglot-events-buffer-size to 0. * You're using some other greedy extension in Emacs, be sure to conduct any tests using a controlled `emacs -Q` run. As usual, the only way to tell is to conduct measurements, so be sure to send some `M-x profiler-report` output this way. FWIW's here's a typical profile from 2-3 minutes editing in a C++ project. As you can see Eglot's code isn't even on the map. But c++-mode's fontification code is, though not in a show-stopping way. c++-mode is notoriously complex and slow (and buggy), maybe someday we'll get to replace it with something better. João

So:
eglot-send-change-idle-time 0.5
eldoc-idle-delay 0.5 (but I have Documentation on Hover turned off, so I think this doesn't apply?)
company-idle-delay 0.2 (but if I'm honest, this isn't working great; I might turn it into an explicit completion for C++ rather than pausing)

To be clear, my current hypothesis is that it is indeed clangd; I saw an 85000+ms response time on a file at some point yesterday, and unless eglot spends a lot of time hitching up for other people on smaller projects, it seems irrational to believe that it's the root cause of my problems now.

**Reply by vincentjgoh** [reply]
So I set my CLANGD_FLAGS environment variable to --enable-config to ensure that my config is indeed being picked up, and the stderr output confirms it. (Actually, it doesn't say anything about it. It confirms that if I deliberately introduce errors it will find them and spit that out.)

I also changed the events buffer size to 5000. 

I tried an initial run with the profiler and completing my variables all went really smoothly and I had a profile devoid of anything interesting. Maybe my problems were solved! I saved the file, and that also went smoothly. I stopped the profiler, removed the line that I typed, and saved again, and THAT'S when a big stall hit me. So I started the profiler again, and tried typing out exactly the same line as before. THIS time, autocomplete choked and actually timed out. I tried again, and it worked after a delay. I finished the line and saved, no problem. I removed the line and saved, emacs hung so hard that Windows' spinning wheel showed up.

  1,958,132,691  63% - command-execute
  1,889,305,220  61%  - funcall-interactively
  1,859,026,763  60%   - save-buffer
  1,859,026,763  60%    - basic-save-buffer
  1,858,646,825  60%     - run-hooks
  1,856,209,743  59%      - eglot--signal-textDocument/didSave
  1,856,014,573  59%       - jsonrpc-notify
  1,856,014,573  59%        - jsonrpc-connection-send
  1,856,014,573  59%         - apply
  1,856,014,573  59%          + #<compiled -0x1ae40cf16d13c42b>
        105,954   0%       + eglot--signal-textDocument/didChange
         21,816   0%       + eglot--TextDocumentIdentifier
      2,389,032   0%      + clang-format+-before-save
         22,072   0%      + doom-modeline-update-buffer-file-name
         12,584   0%      + eglot--signal-textDocument/willSave
          4,136   0%        delete-trailing-whitespace
          3,042   0%      + flymake-after-save-hook
          1,056   0%      + coding-system-hook
        358,342   0%     + basic-save-buffer-1
         21,596   0%     + delete-auto-save-file-if-necessary
     14,766,672   0%   + isearch-repeat-backward
      4,291,923   0%   + self-insert-command
      3,509,991   0%   + execute-extended-command
      3,265,414   0%   + isearch-printing-char
      1,538,654   0%   + company-complete-selection
        703,477   0%   + c-electric-colon
        506,637   0%   + c-electric-semi&comma
        401,090   0%   + c-electric-paren
        225,953   0%   + c-electric-backspace
        223,506   0%   + scroll-down-command
        207,809   0%   + previous-line
        151,769   0%   + kill-region
        129,760   0%   + newline
        108,436   0%   + isearch-backward
        105,308   0%   + kill-line
         13,728   0%   + handle-focus-out
         11,616   0%   + handle-focus-in
          4,224   0%   + handle-switch-frame
          3,168   0%   + mouse-drag-region
          2,112   0%   + mouse-set-point
          1,224   0%   + set-mark-command
            168   0%   + end-of-buffer
     68,826,415   2%  + byte-code
    932,378,256  30% - timer-event-handler
    932,336,680  30%  - apply
    931,890,546  30%   - company-idle-begin
    930,200,987  30%    - company-auto-begin
    930,200,105  30%     - company--perform
    930,200,105  30%      - company--begin-new
    930,047,457  30%       - company-calculate-candidates
    930,041,777  30%        - company--fetch-candidates
    930,041,777  30%         - company-call-backend-raw
    930,026,710  30%          - apply
    930,026,710  30%           - company-capf
    930,001,022  30%            - company-capf--candidates
    929,998,910  30%             - completion-all-completions
    929,998,910  30%              - apply
    929,998,910  30%               + #<subr completion-all-completions>
         10,067   0%            error
          4,624   0%        + run-with-timer
          1,056   0%          company--preprocess-candidates
        148,456   0%       + company-call-backend
      1,689,559   0%    + company-post-command
        262,378   0%   + #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_13>
         52,800   0%     #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_12>
         49,544   0%   + #<compiled -0x10f1396901e06a59>
         33,052   0%   + #<compiled -0x82dfdffed84323a>
         25,344   0%   + isearch-lazy-highlight-start
         17,736   0%   + #<compiled -0x16fae5de3513f695>
          3,168   0%   + blink-cursor-start
          1,056   0%   + auto-revert-buffers
         31,680   0%  + timer-activate-when-idle
          4,416   0%  + timer-activate
          3,608   0%    timer-inc-time
    177,438,091   5% + redisplay_internal (C function)
     15,031,873   0% + jsonrpc--process-filter
     10,978,204   0% + internal-default-process-filter
        370,124   0% + mic-paren-command-hook
         57,962   0% + company-post-command
         15,840   0% + jit-lock--antiblink-post-command
         15,840   0% + which-key--hide-popup
         12,472   0% + #<compiled 0x153dba778db7467b>
          8,024   0% + eldoc-pre-command-refresh-echo-area
          7,416   0% + #<compiled -0x13113ffcc5308b10>
          6,336   0% + blink-cursor-check
          3,456   0% + eldoc-schedule-timer
          3,168   0%   internal-echo-keystrokes-prefix
          3,168   0% + substitute-command-keys
          1,528   0% + tooltip-show-help
          1,128   0% + mouse--click-1-maybe-follows-link
            631   0%   isearch-pre-command-hook
              0   0%   ...

**Reply by joaotavora** [reply]
I just want to emphasize that even if clangd or some external process takes several seconds to answer, or even hangs, that should NOT, in principle, produce the sluggishness you reported earlier.  It simply means that you'll be deprived of LSP feedback like hover information, completions, etc for those periods.

> company-idle-delay 0.2 (but if I'm honest, this isn't working great; I might turn it into an explicit completion for C++ rather than pausing)

This means that company will do a completion request to LSP 0.2 seconds after you type a character, and then likely discard the results of that request very soon.  In my opinion, you should use either "explicit" (the easiest one is the built-in `C-M-i`) or adjust that frequency, which is a frequency similar to a normal typing speed. 

> eldoc-idle-delay 0.5 (but I have Documentation on Hover turned off, so I think this doesn't apply?)

Yeah, if you turned those off it shouldn't do any LSP communication specifically because of eldoc.

**Reply by joaotavora** [reply]
Please expand this line

```
1,856,014,573 59% + #<compiled -0x1ae40cf16d13c42b>
```

The biggest chunk of Elisp time is being spend sending the context of the file to the server after save, it seems.  Saving a buffer however is an occasional action, not like typing characters into a buffer.

Please be very clear about what actions exactly you performed between the time of `M-x profiler-start` and the time of `M-x profiler-stop`.  

Please also try with `eglot-events-buffers-size` set to 0.

**Reply by vincentjgoh** [reply]
I'll do that all presently, though I'd like to mention that I had to edit some deep engine files in the project, and my clangd is currently taking up 75% of my CPU and 42GB of memory despite the fact that I'm not actively editing anything anymore. It briefly peaked its memory usage at 50GB, but the CPU usage was down to 25% at that point.

**Reply by vincentjgoh** [reply]
```
  1,958,132,691  63% - command-execute
  1,889,305,220  61%  - funcall-interactively
  1,859,026,763  60%   - save-buffer
  1,859,026,763  60%    - basic-save-buffer
  1,858,646,825  60%     - run-hooks
  1,856,209,743  59%      - eglot--signal-textDocument/didSave
  1,856,014,573  59%       - jsonrpc-notify
  1,856,014,573  59%        - jsonrpc-connection-send
  1,856,014,573  59%         - apply
  1,856,014,573  59%          - #<compiled -0x1ae40cf16d13c42b>
     36,271,081   1%           - process-send-string
     36,193,271   1%            - internal-default-process-filter
     27,361,383   0%             - #<compiled 0x1fa8af6bdbd6a339>
          1,649   0%                line-beginning-position
         64,793   0%            - jsonrpc--process-filter
         61,025   0%             - jsonrpc-connection-receive
         36,889   0%              - #<compiled -0xd98af4b9f55e46c>
         36,889   0%               - apply
         36,889   0%                - eglot-handle-notification
         36,889   0%                 - apply
         36,889   0%                  - #<compiled 0xfe0f1c02cdbdad7>
         23,888   0%                   + find-buffer-visiting
         11,840   0%                   + eglot--uri-to-path
          1,056   0%                   + eglot--check-object
             85   0%                     #<compiled -0x9de98e34fd5d35d>
             20   0%                   + eglot--range-region
         24,136   0%              + jsonrpc--log-event
          1,056   0%               #<compiled -0x1de56117c0c1b097>
          9,849   0%            + timer-event-handler
        589,860   0%           - jsonrpc--log-event
        378,616   0%            - pp-to-string
         12,376   0%             - pp-buffer
         12,376   0%              + indent-sexp
          1,056   0%               #<compiled -0x1de5613dbcf5f097>
        139,448   0%             jsonrpc--json-encode
        105,954   0%       - eglot--signal-textDocument/didChange
         84,666   0%        - jsonrpc-notify
         84,666   0%         - jsonrpc-connection-send
         84,666   0%          - apply
         84,666   0%           - #<compiled -0x1ae40cf16d13c42b>
         81,504   0%            + jsonrpc--log-event
         20,232   0%        - eglot--VersionedTextDocumentIdentifier
         20,232   0%         - eglot--TextDocumentIdentifier
         20,232   0%          - eglot--path-to-uri
         12,376   0%           - url-hexify-string
         12,376   0%            + mapconcat
          7,344   0%           + file-truename
         21,816   0%       + eglot--TextDocumentIdentifier
```

**Reply by vincentjgoh** [reply]
Actually, I'll have to do those actions later; I had to kill clangd; my compile couldn't allocate enough memory to use any more than 1 process instead of the regular 8.

Thanks so much for all your help so far!

**Reply by vincentjgoh** [reply]
In the end, I don't think it had anything to do with Eglot OR clang. I stripped my emacs down to the bare bones and built it up bit by bit, and the problem is gone. It was honestly VSCode that tipped me off that something was amiss, because it was incredibly fast using the exact same clang index.

**Comment by joaotavora** [comment, marked as answer]
> I don't think it had anything to do with Eglot OR clang. I stripped my emacs down to the bare bones and built it up bit by bit, and the problem is gone. 

No problem. A configuration issue, as is frequent.
